File: src/ai/node_watcher/src/Node.cpp
#include "node_watcher/Node.hpp"

Node::Node(string nodename, string package)
    : NodeStatusHandler(), nodename(nodename) {
    
    this->nodepath = NodeStatusHandler::makeNodePath(nodename, package);

    // Wait for service to send init signal
    setNodeStatus(NodeStatus::NODE_INIT);

    // Set wait callback
    this->setWaitCallback(boost::bind(&Node::onWaitingResult, this, ::_1));
}

Node::~Node() {
    setNodeStatus(NodeStatus::NODE_DESTROYED);
}


void Node::onAwaitResponse(const AwaitNodesResult& msg) {
    if (msg.request_code == this->waitRequestCode) {
        ROS_INFO_STREAM("Node " << nodename << " is done waiting for nodes");
        
        this->onWaitingResult(msg.success);
    }
}

/**
 * Set current node status 
 */
void Node::setNodeStatus(int state_code, int errorCode /*= 0*/) {
    // update self status
    this->status.state_code = state_code;
    this->status.error_code = errorCode;

    // update remotely
    NodeStatusHandler::setNodeStatus(nodepath, state_code, errorCode);
}

/**
 * Retrieve current node status
 */
NodeStatus Node::getNodeStatus(bool remote /*= false*/) {
    if (remote) {
        return NodeStatusHandler::getNodeStatus(this->nodepath);
    } else {
        return this->status;
    }
}

/**
 * Wait for nodes described in node's package requirements.txt file
 */
void Node::waitForNodes(int timeout) {
    string filename = ros::package::getPath(this->nodename) + "/requirements.txt";
    this->requireFromFile(filename);
    NodeStatusHandler::waitForNodes(timeout);
}

void Node::waitForNodes(int timeout, bool useFile) {
    if (useFile) {
        this->waitForNodes(timeout);
    } else {
        NodeStatusHandler::waitForNodes(timeout);
    }
}

File: src/ai/node_watcher/include/node_watcher/node.py
import rospy
import rospkg

from node_watcher.node_status_handler import NodeStatusHandler

from ai_msgs.msg import AwaitNodesResult, Topics, NodeStatus
from std_msgs.msg import Int32

from abc import ABC, abstractmethod

class Node (NodeStatusHandler, ABC):
	'''
		Class describing a generic ROS node, registered to
		the node_watcher_node.
	'''

	def __init__(self, nodename: str, package: str):
		super().__init__()

		self.nodename = nodename
		self.package = package

		self._status = NodeStatus()
		self.set_status(NodeStatus.INIT)

		self.set_wait_callback(self.on_waiting_result)

	def _on_await_response(self, msg: AwaitNodesResult):
		if msg.request_code == self._wait_request_code:
			rospy.loginfo("Node {} is done waiting for nodes".format(self.nodename))
			self._wait_callback(msg.success)

	def on_waiting_result(self, success: bool):
		pass 

	# Status setter
	def set_status(self, state_code: int, error_code: int = 0):
		# Save status
		self._status.state_code = state_code
		self._status.error_code = error_code

		# Update remotely
		super().set_node_status(
			self.nodename, self.package, state_code, error_code
		)

	# add self status getter
	def get_status(self, remote: bool = False):
		if remote:
			return super().get_node_status(self.nodename, self.package)
		else:
			return self._status

	def wait_for_nodes(self, timeout: int, use_pkg_file: bool = False):
		# If we use the package "requirements.txt" file
		if use_pkg_file:
			# We add requirements to it
			filename = rospkg.RosPack().get_path(self.nodename) + "/requirements.txt"
			self.require_from_file(filename)
		
		# Otherwise just wait
		super().wait_for_nodes(timeout)

